lib/std/os.zig

```diff
@@ -5021,6 +5022,89 @@ pub fn copy_file_range(fd_in: fd_t, off_in: u64, fd_out: fd_t, off_out: u64, len
     return pwrite(fd_out, buf[0..amt_read], off_out);
 }
 
+var has_copy_file_range_syscall = std.atomic.Int(u1).init(1);
+
+pub const CopyFileOptions = struct {};
+
+pub const CopyFileError = error{
+    BadFileHandle,
+    SystemResources,
+    FileTooBig,
+    InputOutput,
+    IsDir,
+    OutOfMemory,
+    NoSpaceLeft,
+    Unseekable,
+    PermissionDenied,
+    FileBusy,
+} || FStatError || SendFileError;
+
+/// Transfer all the data between two file descriptors in the most efficient way.
+/// No metadata is transferred over.
+pub fn copy_file(fd_in: fd_t, fd_out: fd_t, options: CopyFileOptions) CopyFileError!void {
+    if (comptime std.Target.current.isDarwin()) {
+        const rc = system.fcopyfile(fd_in, fd_out, null, system.COPYFILE_DATA);
+        switch (errno(rc)) {
+            0 => return,
+            EINVAL => unreachable,
+            ENOMEM => return error.SystemResources,
+            // The source file was not a directory, symbolic link, or regular file.
+            // Try with the fallback path before giving up.
+            ENOTSUP => {},
+            else => |err| return unexpectedErrno(err),
+        }
+    }
+
+    if (std.Target.current.os.tag == .linux) {
+        // Try copy_file_range first as that works at the FS level and is the
+        // most efficient method (if available).
+        if (has_copy_file_range_syscall.get() != 0) {
```

Comment:
This should use the target version range to determine which state we are in:
 * the target OS version range guarantees kernel will have `copy_file_range`. Skip the runtime check.
 * the target OS version range guarantees kernel will not have `copy_file_range`. Skip the runtime check.
 * the target OS version range spans a kernel version that has the syscall and a version that does not have the syscall. Do the runtime check.

Reply:
Sure thing, it's enough to preload `has_copy_file_range_syscall` with 0 if the version check fails.